search/src/Filtering.cpp

```diff
@@ -46,6 +46,98 @@ std::vector<int> sigmaGFilteredIndices(const std::vector<float>& values, float s
     return result;
 }
 
+std::vector<int> clippedAverageFilteredIndices(const std::vector<float>& psi_curve,
+                                               const std::vector<float>& phi_curve,
+                                               int num_clipped, int n_sigma, float lower_lh_limit) {
+    /*
+     * This function applies a clipped median filter to a set of likelihood values. The largest
+     * likelihood values (N=num_clipped) are eliminated if they are more than n_sigma*st_dev
+     * away from the median, which is calculated excluding the largest values.
+     *
+     * Arguments:
+     *     psi_curve - The psi values.
+     *     phi_curve - The phi values.
+     *     num_clipped - The number of likelihood values to consider eliminating. Only
+     *                   considers the largest N=num_clipped values.
+     *     n_sigma - The number of standard deviations away from the median that the largest
+     *               likelihood values (N=num_clipped) must be in order to be eliminated.
+     *     lower_lh_limit - Likelihood values lower than lower_lh_limit are automatically
+     *                      eliminated from consideration.
+     *
+     * Returns: The indices of psi & phi that pass the filtering.
+     */
+    const int num_times = psi_curve.size();
+    assert (num_times == phi_curve.size());
+
+    // Compute the likelihood for each index.
+    std::vector<std::pair<float, int> > scores;
+    for (int t = 0; t < num_times; ++t) {
+        float val = 0.0;
+        if ((psi_curve[t] == NO_DATA) || (phi_curve[t] == NO_DATA)) {
+            val = NO_DATA;
+        } else if (phi_curve[t] > 0.0) {
+            val = psi_curve[t] / sqrt(phi_curve[t]);
+        }
+        scores.push_back(std::make_pair(val, t));
+    }
+    sort(scores.begin(), scores.end());
+
+    // Find all the indices where:
+    // 1) LH > lower_lh_limit
+    // 2) They do not have one of the top 'num_clipped' likelihoods.
+    // 3) LH != NO_DATA
+    int first_valid = 0;
+    int last_valid = 0;
+    double lh_sum_good = 0.0;
+    for (int t = 0; t < num_times - num_clipped; ++t) {
```

Comment:
did you consider using an iterator here? I am curious if it makes it any faster or if the compiler optimizes it to be equivalent.

Reply:
I didn't benchmark that particular change.